docs/guillotine.adoc

```diff
@@ -0,0 +1,1349 @@
+
+    return {
+        contentType: 'application/json',
+        body: executeQuery(body.query, body.variables),
+        status: 200
+    };
+};
+----
+
+<1> At the core is the function `executeQuery`. Here, a guillotine `SCHEMA` definition is combined with a graphQL `query` string and an optional `variables` object. These are used with XP's graphQL library to `execute` the query. The result, a JSON object, is returned.
+<2> `executeQuery` is exposed and directly usable from an XP controller. That's what we'll do next.
+<3> a `post` function is also included for receiving POST requests from outside, e.g. a browser. If these requests contain a query string, it's executed with `executeQuery` above, and the result is returned in a response: basically a complete guillotine API endpoint for your webapp.
+
+NOTE: This endpoint is *disabled by default* in the starter, to encourage developers to consider security aspects before using it. We'll get back to that, and activate it, <<#expose_api, later>>.
+
+{zwsp} +
+
+The second included helper, <<#guillotine_request_helper, guillotineRequest.es6>>, is a `fetch` wrapper to simplify guillotine requests at the frontend. We'll take a look at that later.
+
+{zwsp} +
+
+=== Domain-specific helper for listing movies
+
+In order to make requests for a list of movies below a container item in the content hierarchy, we'll need a specific guillotine query string, as well as functionality to adapt the resulting data into the proper props structure for the react components.
+
+And by using the same code on the frontend and backend, for this too, we gain a bit of isomorphism (the predictability of a singe source of truth, in short). So we'll make *a module with custom helper functionality* for our use case, and import from it in both places.
+
+Let's go ahead an write this:
+
+.headless/helpers/movieListRequests.es6:
+[source,javascript,options="nowrap"]
+----
+// Used by both backend and frontend (the movie-list part controller, and react4xp/entries/MovieList.jsx)
+
+             <!--1-->
+export const buildQueryListMovies = () => `
+query(
+    $first:Int!,
+    $offset:Int!,
+    $sort:String!,
+    $parentPathQuery:String!
+) {
+  guillotine {
+    query(
+        contentTypes: ["com.enonic.app.react4xp:movie"],
+        query: $parentPathQuery,
+        first: $first,
+        offset: $offset,
+        sort: $sort
+    ) {
+      ... on com_enonic_app_react4xp_Movie {
+        _id
+        displayName
+        data {
+          year
+          description
+          actor
+          image {
+            ... on media_Image {
+              imageUrl(type: absolute, scale: "width(300)")
+            }
+          }
+        }
+      }
+    }
+  }
+}`;
+
+             <!--2-->
+export const buildParentPathQuery = (parentPath) => `_parentPath = '/content${parentPath}'`;
+
+
+// Not using util-lib, to ensure usability on frontend
+const forceArray = maybeArray => Array.isArray(maybeArray)
```

Comment:
Double ternary 😕

Reply:
If you write your own forceArray just use that everywhere. No need to include lib-util then.

Reply:
Good points.

Reply:
I'll need to restructure a little bit of code then, and subsequently the docs. Suggestion: we release the docs without this improvement, and create a polish-task for it?

Reply:
❓

Reply:
- [ ] https://github.com/enonic/doc-react4xp/issues/8